solana/solana-ibc/programs/solana-ibc/src/chain.rs

```diff
@@ -0,0 +1,162 @@
+use anchor_lang::prelude::*;
+use anchor_lang::solana_program;
+pub use blockchain::Config;
+
+use crate::error::Error;
+use crate::storage;
+
+type Result<T = (), E = anchor_lang::error::Error> = core::result::Result<T, E>;
+
+pub type Epoch = blockchain::Epoch<PubKey>;
+pub type Block = blockchain::Block<PubKey>;
+pub type Manager = blockchain::ChainManager<PubKey>;
+pub use crate::ed25519::{PubKey, Signature, Verifier};
+
+/// Guest blockchain data held in Solana account.
+#[account]
+pub struct ChainData {
+    inner: Option<ChainInner>,
+}
+
+impl ChainData {
+    /// Initialises a new guest blockchain with given configuration and genesis
+    /// epoch.
+    ///
+    /// Fails if the chain is already initialised.
+    pub fn initialise(
+        &mut self,
+        trie: &mut storage::AccountTrie,
+        config: Config,
+        genesis_epoch: Epoch,
+    ) -> Result {
+        let (height, timestamp) = host_head()?;
+        let genesis = Block::generate_genesis(
+            1.into(),
+            height,
+            timestamp,
+            trie.hash().clone(),
+            genesis_epoch,
+        )
+        .map_err(|err| Error::Internal(err.into()))?;
+        let manager = Manager::new(config, genesis).map_err(Error::from)?;
+        if self.inner.is_some() {
+            return Err(Error::ChainAlreadyInitialised.into());
+        }
+        let last_check_height = manager.head().1.host_height;
+        self.inner = Some(ChainInner { last_check_height, manager });
+        Ok(())
```

Comment:
maybe we would like to submit an event here? Stating that the chain has started and the hash is 0x*** and any other relevant data?
No info about the val set?

Reply:
D’oh, yes, this got lost in rebases. I’ll fix it in a moment.